File: python-algorithm/ds_5_tree/tree_2_traversal/L589.cpp


#include <iostream>
#include <vector>
using namespace std;

// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};

class Solution {
public:
    vector<int> preorder(Node* root) {
        vector<int> res;
        mypreorder(root, res);
        return res;
    }

    //前序遍历打印，递归实现
    void mypreorder(Node* node , vector<int> &res){
        if(node == NULL)  //递归终止条件
            return;
        res.push_back(node->val) ;
        for(auto point:node->children)
            mypreorder(point,res);
    }
};

//非递归写法
#include <stack>
class Solution1 {
public:
    vector<int> preorder(Node* root) {
        vector<int> res;
        if(root == NULL) return res;
        stack<Node*> s;
        s.push(root);
        while(!s.empty()){
            Node* cur = s.top();
            s.pop();
            if(cur == NULL) continue;
            res.push_back(cur->val);
            for(int i=cur->children.size()-1; i>=0; --i)
                s.push(cur->children[i]);
        }
        return res;
    }
};

File: python-algorithm/ds_5_tree/tree_2_traversal/L589.py
# -*- coding: utf-8 -*-


class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        res = []
        if not root:
            return res
        self.mypre(root, res)
        return res

    def mypre(self, root, res):
        if not root:
            return
        res.append(root.val)
        for node in root.children:
            self.mypre(node, res)

class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        res = []
        if not root:
            return res
        stk = []
        stk.append([root, "go"])
        while stk != []:
            cur = stk.pop(-1)
            if cur[1] == "print":
                res.append(cur[0].val)
            else:
                for i in range(len(cur[0].children)-1, -1, -1):
                    stk.append([cur[0].children[i], "go"])
                stk.append([cur[0], "print"])
        return res